server/game/core/ongoingEffect/OngoingEffectEngine.ts

```diff
@@ -152,6 +152,10 @@ export class OngoingEffectEngine {
         this.effectsChangedSinceLastCheck = this.unapplyAndRemove((effect) => effect.duration === Duration.UntilEndOfRound);
     }
 
+    private onActionEnded() {
```

Comment:
I may not have time to do a full review right now but wanted to give one piece of top-level feedback. We might want to couple the cost adjuster a little more tightly to the specific ability that's being resolved. I worry that if we activate it for the duration of the action we still may run into issues like e.g. if someone triggers the ability but then hits the cancel button at the target stage to revert back to action selection.

An approach that might be useful both as a future feature and for general bug protection would be to make it so that the cost adjust can be configured to only match the specific play card action that we're intending it for. Haven't thought about this deeply but maybe a way to solve this is that cost adjusters don't have to be bound to the player, but could also be provided from the ability itself? That seems like a fairly elegant way to provide these one-off adjusters that are tied directly to the executing action. Could just add it as a new constructor parameter / property for PlayCardAction.

Reply:
I think that is a good idea. I finally figured out where to put things for the implementation of it, but I ran into a really strange error: "AbilityContext is not a constructor" when it's called in PlayerAction. I didn't change PlayerAction or AbilityContext at all so I'm not sure what's wrong.